scipy/optimize/_nnls.py

```diff
@@ -72,14 +88,73 @@ def nnls(A, b, maxiter=None):
                 "Incompatible dimensions. The first dimension of " +
                 f"A is {m}, while the shape of b is {(b.shape[0], )}")
 
-    maxiter = -1 if maxiter is None else int(maxiter)
-
-    w = zeros((n,), dtype=double)
-    zz = zeros((m,), dtype=double)
-    index = zeros((n,), dtype=int)
-
-    x, rnorm, mode = __nnls.nnls(A, m, n, b, w, zz, index, maxiter)
+    x, rnorm, mode = _nnls(A, b, maxiter, tol=atol)
     if mode != 1:
-        raise RuntimeError("too many iterations")
+        raise RuntimeError("Maximum number of iterations reached.")
 
     return x, rnorm
+
+
+def _nnls(A, b, maxiter=None, tol=None):
+    """
+    This is a single RHS algorithm from ref [2] above. For multiple RHS
+    support, the algorithm is given in  :doi:`10.1002/cem.889`
+    """
+    m, n = A.shape
+
+    AtA = A.T @ A
```

Comment:
Condition number? Looks like an SVD use case.

Reply:
Didn't understand your comment. Where do you want to use SVD?

Reply:
You are constructing `A.T @ A`, which has the condition number a square of the condition number of `A`. And you're using it as a l.h.s. of a linear solve routine.

Reply:
True, albeit partial solves. The algorithm goes through submatrices of `AtA` to only deal with active constraints hence the `A[P, P]` dance.